### ISSUE TITLE
Foreign syntax doesn't recognize tables in other modules

### ISSUE THREAD

**Issue by parsonsmatt** [yesodweb/persistent#1073]
Suppose you have the following modules:

```haskell
-- Schema/Foo.hs
module Schema.Foo where

mkPersist $(persistFileWith "foo.persistentmodels")

-- foo.persistentmodels
Foo
    name Text
    Primary name

-- Schema/Bar.hs

module Schema.Bar where

import Schema.Foo

mkPersist $(persistFileWith "bar.persistentmodels")

-- bar.persistentmodels
Bar
    foo Text
    Foreign Foo fk_bar_to_foo foo
```

This code will currently fail, because the `HaskellName "Foo"` is not present in the `EmbedEntityMap`, because it was not parsed from the same block of text.

I think there needs to be some way to lookup the entity *without* requiring it be present in the map. Attempting to call `entityDef (Nothing :: Maybe $(Foo))` instead of relying on what's in the `Map` might work.

**Comment by korayal** [comment]
It seems this issue still exists on persistent-2.11 where setting foreign key constrains on a non-primary key is introduced. It would be unfortunate to not be able to use it on a split-file setup.

Are there any solutions other than combining these separated model files back? looks like we can do `parent FooId` using the primary key of the Foo table, but not this:

``` haskell
-- Schema/Foo.hs
module Schema.Foo where

mkPersist $(persistFileWith "foo.persistentmodels")

-- foo.persistentmodels
Foo
    name Text
    UniqueName name

-- Schema/Bar.hs

module Schema.Bar where

import Schema.Foo

mkPersist $(persistFileWith "bar.persistentmodels")

-- bar.persistentmodels
Bar
    foo Text
    parent Text
    Foreign Foo fk_foo_name parent References name
```

**Comment by parsonsmatt** [comment]
Good catch, thanks.

That error is getting thrown here: [Database.Persist.Quasi.fixForeignKeysAll](https://github.com/yesodweb/persistent/blob/9e6d88d767035d3e62252666ab3083ab576257bf/persistent/Database/Persist/Quasi.hs#L772-L778). We use the `pent :: EntityDef` to get the `DBName` for the entity, as well as the length of fields on the primary key (unnecessary and also a bug that probably slipped in).

`fixForeignKeysAll` is called [here](https://github.com/yesodweb/persistent/blob/9e6d88d767035d3e62252666ab3083ab576257bf/persistent/Database/Persist/Quasi.hs#L644-L657), and `parse` is called in [`parseReferences`](https://github.com/yesodweb/persistent/blob/9e6d88d767035d3e62252666ab3083ab576257bf/persistent-template/Database/Persist/TH.hs#L247), which is used in the *quasiquoter*, not in the `mkPersist` code, which would be able to resolve this by performing some other lookup.

`parseReferences` *is* in `Q`, which means that we could translate these functions to `Q`. But then we're not returning `EntityDef`, we're returning `Q Exp`, though we could use the typed stuff and return `Q (TExp EntityDef)`.

The `UnboundForeignDef` is created in `takeForeign`, which just puts a dummy default. So in `fixForeignKeysAll`, it's trying to overwrite the bad `DBName` with a good `DBName` from the `EntityDef`. We could simply defer that check until the `liftAndFixKeys` step that occurs in the TemplateHaskell code.

tl;dr:

1. We need an `EntityDef` for the referenced `Entity` in order to know the right `DBName` for the referenced table.
2. We don't necessarily have an `EntityDef` at parse-time, but we will at compile-time.
3. We need to defer the check of the validity of the foreign key until we are doing `liftAndFixKeys`, which takes place in the `mkPersist` stage of things. That will allow us to splice in the `EntityDef` that has been "fixed" already.

I really want to refactor the QQ and TH machinery to use more special-cased datatypes. The `QQ` shouldn't be outputting `EntityDef` since it can't possibly know. `UnboundEntityDef` is the right approach, but that messes up also by containing an `EntityDef` that needs to be "fixed." A better approach would be a separate type, and then the necessary function would be something like `translate :: UnboundEntityDef -> Q (TExp EntityDef)`.

**Comment by jkachmar** [comment]
Just as a small addition to this, but it looks like `parent FooId` (in @korayal's example) has a more subtle problem.

It compiles just fine, but the migration that `persistent` generates will fail to include foreign key constraints.

e.g. Following the original example
```haskell
-- bar.persistentmodels
Bar
    foo Text
    parent FooId
```
...will produce a table that contains `parent` with `NOT NULL` constraints and a `BIGINT` type, but no foreign key constraints at all.

**Comment by parsonsmatt** [comment]
I'm going to close this out, since we have a work-around.

See #1241 for more details.

tl;dr:

```haskell
import Foo
import Bar

mkPersist sqlSettings . mappend $(discoverEntities) [persistLowerCase| ... |]
```
